students/jason_jenkins/lesson03/mailroom.py

```diff
@@ -9,57 +9,151 @@
 
 # Global Variables
 run_program = True
+donor_list = []
 
 
 def send_thanks():
+    """
+    Method used to probt donor name or list out donors
+    """
+
+    global donor_list
     response = ""
+
     while(True):
-        print("You want to send a thank you.")
-        response = input('Input persons full name or view list my inputing "list": ')
+        response = input('Input donors full name, "list", or "exit": ')
 
-        if(response == "list"):
-            print("Print the List of names")
-        elif(response != ""):
-            print(response)
+        if(response.lower() == "exit"):
             break
+        elif response.lower() == "list":
+            print_donor_list()
+            print()
         else:
             break
 
+    if(response != "exit"):
+        for donor in donor_list:
+            if donor[0] == response:
+                donate(donor)
+                break
+        else:
+            new_donor = [response]
+            donate(new_donor)
+            donor_list.append(new_donor)
+
+
+def thank_you_email(donor):
+    """
+    Thank donor for donation
+    """
+
+    print(f"Thank you {donor[0]} for your donation.")
+
+
+def print_donor_list():
+    """
+    Print out the donor list
+    """
+
+    global donor_list
+
+    print("List of donors")
+    print("--------------")
+
+    for donor in donor_list:
+        print(donor[0])
+
+
+def donate(donor):
+    """
+    Prompt donor to donate
+    """
+
+    response = float(input('Input amount to donate or "0" to exit: '))
+    if(response != 0):
+        donor.append(response)
+        thank_you_email(donor)
+
 
-def donate():
-    print("add donation")
+def sort_by_total(tmp_donor_list):
+    """
+    Used to sort donor list by total donations
+    """
+
+    return sum(tmp_donor_list[1:])
 
 
 def create_report():
-    print("create_report")
+    """
+    Create a table like view of donors
+    Includes donor name, donation total, total gifts, average gift amount
 
+    Sorted by donation total
+    """
+
+    global donor_list
+
+    donor_list.sort(key=sort_by_total, reverse=True)
+
+    print(f"{'Donor Name':30}|{'Total Given':^16}|", end='')
+    print(f"{'Num Gifts':^14}|{'Average Gift':^16}")
+    print(f"{'-'*79}")
+
+    for donor in donor_list:
+        donor_name = donor[0]
+        donor_total = sum(donor[1:])
+        donar_count = len(donor) - 1
+        donor_ave = 0
+
+        if donar_count != 0:
+            donor_ave = donor_total / donar_count
+
+        donor_output = f"{donor_name:30}"
+        donor_output += f" ${donor_total:15.2f}"
+        donor_output += f"{donar_count:15}"
+        donor_output += f" ${donor_ave:15.2f}"
+        print(donor_output)
+
+
+def quit_program():
+    """
+    Method used to quit the program
+    """
 
-def startup_prompt():
     global run_program
+    run_program = False
```

Comment:
this should utilize sys.exit() (this will be needed especially in mailroom version 2)

see tutorial: https://uwpce-pythoncert.github.io/PythonCertDevel/exercises/mailroom_tutorial.html